// analysis/db/analysis.js

var sql = require('fusion/sql');

exports.check = function() {
  // Determine if all jobs have been completed
  var nullDate = '0000-00-00 00:00:00';
  var res = sql.query('SELECT * FROM worklist WHERE CheckIn=?',[nullDate]);
  return res.length == 0;
}

exports.getCurrentModel = function() {
  var res = sql.query('SELECT drugid FROM worklist WHERE 1 LIMIT 1');
  return res.length > 0 ? res[0].drugid : 0;
}

exports.getCurrentModelName = function() {
  var res = sql.query('SELECT name FROM drugs WHERE id=?',[exports.getCurrentModel()]);
  return res.length > 0 ? res[0].name : '';
}

exports.archive = function() {
  // Determine the current drugID
  var res = sql.query('SELECT Drug FROM results WHERE 1 LIMIT 1');
  if (res.length > 0) {
    var drugID = res[0].drug;
    res = sql.query('SELECT * FROM drugs WHERE id=? LIMIT 1',[drugID]);
    if (res.length > 0) {
      var drugName = res[0].name;
      sql.query('RENAME TABLE results TO results_'+drugName);
      sql.query('RENAME TABLE worklist TO worklist_'+drugName);
      sql.query('UPDATE drugs SET completed=1 WHERE id=? LIMIT 1',drugID);
    }
  }
}

exports.getAvailableModels = function() {
  return sql.query('SELECT id,name FROM drugs WHERE completed=0');
}

exports.tableExists = function(table) {
  var res = sql.query('SHOW TABLES LIKE \''+table+'\'');
  return res.length > 0;
}

exports.checkTables = function() {
  return !exports.tableExists('worklist') && !exports.tableExists('results');
}

exports.hasResults = function() {
  if (exports.tableExists('results')) {
    var res = sql.get('SELECT COUNT(*) as c FROM results');
    return res.c > 0;
  } else return false;
}

exports.createNewAnalysis = function(drugID) {
  sql.query('CREATE TABLE results LIKE results_template');
  sql.query('CREATE TABLE worklist LIKE worklist_template');
  sql.query('INSERT worklist SELECT * FROM worklist_template');
  sql.query('UPDATE worklist SET drugid=? WHERE 1',[drugID]);
}

exports.clearWorklist = function() {
  var nullDate = '0000-00-00 00:00:00';
  var res = sql.query('SELECT * FROM results');
  if (res.length == 0) {
    sql.query('UPDATE worklist SET nodeid=0,checkout=?,checkin=?',[nullDate,nullDate]);

    return sql.data('SELECT users.CompID,users.UserID,users.Username,users.Response,users.LastCheckInOut FROM computers LEFT JOIN users ON computers.CompID = users.CompID WHERE Active = 1');
    return true;
  } else return false;
}

exports.hasOutstandingJobs = function() {
  // Determine if all jobs have been completed
  var nullDate = '0000-00-00 00:00:00';
  var res = sql.query('SELECT * FROM worklist WHERE checkout!=? AND checkin=?',[nullDate,nullDate]);
  return res.length > 0;
}

exports.clearDeadJobs = function() {
  var nullDate = '0000-00-00 00:00:00';
  sql.query('UPDATE worklist SET nodeid=0,checkout=? WHERE checkin=?',[nullDate,nullDate]);
}
